### DISCUSSION TITLE
Use a fetched locale

### DISCUSSION THREAD

**Discussion by Sashkan** [amannn/next-intl#892] [Q&A]
From what I understand, this library heavily relies on routing to handle locale, by using the `[locale]` directory. Instead, I'd like to set my locale based on my user settings, which are fetched on app load in my root layout

It is possible to do so ? I tried several things, including passing down directly the locale value to both the `NextIntlClientProvider` and the html tag, but it doesn't work

**Comment by amannn** [comment]
> this library heavily relies on routing to handle locale

Yep, that's correct!

Maybe you could dynamically configure the middleware to pick up the user locale while hiding it from the pathname that is displayed to the user:

```tsx
import createIntlMiddleware from 'next-intl/middleware';
import {NextRequest} from 'next/server';
import {getUserLocale} from './utils';
 
export default async function middleware(request: NextRequest) {
  const userLocale = getUserLocale(request) || 'en';
 
  const handleI18nRouting = createIntlMiddleware({
    locales: ['en', 'de'],
    defaultLocale: userLocale,
    localeDetection: false,
    localePrefix: 'never'
  });
  
  return handleI18nRouting(request);
}

// ...
 
```

(cross-posted this answer to https://github.com/amannn/next-intl/discussions/778#discussioncomment-8607703)

**Reply by amannn** [reply]
I've opened https://github.com/amannn/next-intl/issues/960 to find a better solution for this use case.
